### PR TITLE
Adding fixed calc

### PR DESCRIPTION
**Pull request by ciraelith**
Tasks:

1. Update test - check the last test.
2. Correct typos and file mistakes according to the styleguide.

### CODE REVIEW

File: reverse.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    "-": function (a, b) { return a - b },
+    "*": function (a, b) { return a * b }
+  };
+  var operators = extraOperators ? mergeOptions(baseOperators, extraOperators) :
+    baseOperators;
+
+  if (string === undefined) {
+    console.log("Please put an equation to solve in a form of string");
+  }
+  else {
```
Comment (PinkaminaDianePie) [inline reverse.js]: in case of such `if`, where on one branch you have some error throw/console.log with error etc, and in other branch your main algorithm, it's normal practice to exit from entire function in first branh. here is an example

```javascript
//bad
function(){
  if (!correctEmail){
    console.log('email is not correct');
  } else {
    if (!correctPassword){
      console.log('password is not correct');
    } else {
      if (!hasPermissions){
        console.log('permissions is not correct');
      } else {
        //here is main algorithm. and nesting level can be even deeper!
      }
    }
  }
}

//ok
function(){
  if (!correctEmail){
    console.log('email is not correct');
    return;
  }
  if (!correctPassword){
    console.log('password is not correct');
    return;
  }
  if (!hasPermissions){
    console.log('permissions is not correct');
    return;
  }
  //here is main algorithm
}
```

in second example you have only one level of nesting. and it's not only visual issue, just imagine if you'll need to add one additional check somewhere between. in first example it will be mess with spaces/tabs, in second you just add your code.

File: reverse.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    var arr = string.split(" ");
+    for (var k = 0; k < string.length;  k++) {
+      character = arr[k];
+
+      if (/\d/.test(character)){
+        stack.push(character);
```
Comment (PinkaminaDianePie) [inline reverse.js]: parseInt should be here, not in `else` branch, i mention it in previous review

File: reverseE6.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+const string = process.argv.slice(2)[0];
+var extraOperators = process.argv.slice(2)[1];
```
Comment (PinkaminaDianePie) [inline reverseE6.js]: never use `var` in es6, use only const & let

File: reverseE6.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+function reverseCalc(string, extraOperators) {
+  const stack = [];
+  const baseOperators = {
+    "+": (a, b) => a + b,
+    "-": (a, b) => a - b,
+    "*": (a, b) => a * b,
+  };
+  var operators = Object.assign(baseOperators, extraOperators);
```
Comment (PinkaminaDianePie) [inline reverseE6.js]: You have been mutated baseOperators, since `Object.assign` not returning new object, it merges all provided objects into first one. So in your case `baseOperators` and `operators` will be the same after operation. If you need to merge some objects without mutation of existed, you should provide empty object as first parameter of `Object.assign` : 
```javascript
const operators = Object.assign({}, baseOperators, extraOperators);
```

File: reverseE6.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  if (string === undefined) {
+    console.log("Please put an equation to solve in a form of string with spaces");
+  }
+  else {
+
+    let arr = string.split(" ");
+    for (let k = 0; k < string.length;  k++) {
```
Comment (PinkaminaDianePie) [inline reverseE6.js]: in such case it can be easily replaced by for-of cycle `for (let character of string)`

File: reverse.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
+var string = process.argv.slice(2)[0];
+var extraOperators = process.argv.slice(2)[1];
+
+function mergeOptions(obj1, obj2){
+    var obj3 = {};
+    for (var attrname in obj1) { obj3[attrname] = obj1[attrname]; }
```
Comment (olvnikon) [inline reverse.js]: If it is an inline statement you don't need curly braces.
https://github.com/airbnb/javascript/blob/master/README.md#blocks--braces

File: reverse.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+function reverseCalc5(string, extraOperators) {
+  var stack = [];
+  var character;
+  var baseOperators = {
+    "+": function (a, b) { return a + b },
+    "-": function (a, b) { return a - b },
+    "*": function (a, b) { return a * b }
+  };
+  var operators = extraOperators ? mergeOptions(baseOperators, extraOperators) :
```
Comment (olvnikon) [inline reverse.js]: It is too big for one line. Please update the ternary operator according to the style guide.
https://github.com/airbnb/javascript/blob/master/README.md#comparison--nested-ternaries

File: test.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
+const assert = require('assert');
+const reverseCalc = require('./reverseE6').reverseCalc;
+const reverseCalc5 = require('./reverse').reverseCalc5;
+
+
+// TODO: add tests for extraOperators"
+
```
Comment (PinkaminaDianePie) [inline test.js]: You also need to add tests for empty string and string with wrong characters, e.g. "foobar++". There is also no tests for strings with more than 2 numbers. You need to try cover all edge-cases.

File: reverse.js (hunk 1 above)
Comment (olvnikon) [reply reverse.js]: And one more comment the style. `else` should be on the same line as `}`.
https://github.com/airbnb/javascript/blob/master/README.md#blocks--cuddled-elses

File: reverse.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    baseOperators;
+
+  if (string === undefined) {
```
[the remark this answers is not in the document]
Comment (olvnikon) [reply reverse.js]: It will work in this case because `string` is an argument and it exists anyway. But if there was no `string` variable at all it would through an error like Igor said.
Anyway, it is better to use `!!string` to check if it is empty because `string` can be an empty string or null or smth.
https://github.com/airbnb/javascript/blob/master/README.md#coercion--booleans

File: test.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+describe('reverseCalc', function() {
+    it('should properly add 2+2 during the test', function() {
+      const addition = "22+";
```
Comment (PinkaminaDianePie) [inline test.js]: it seems like `22+` is not working with your code, since you split your input string by whitespaces. are you sure if your run your tests?

File: test.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      const write = "2";
+        assert.equal(reverseCalc(write), 2);
+    });
+    it('should throw an error if there is more than one result', function() {
+      const error = "-2 2+";
+      assert.throws(function () { reverseCalc(error) }, Error, "Error thrown");
+    });
+});
+
+describe('reverseCalc5', function() {
```
Comment (PinkaminaDianePie) [inline test.js]: It's not good to copy-paste tests (like any other code), if you wanna be sure if same test cases should be applied to 2 different methods, you need to wrap them in function and call it for these functions. Your code will be smaller and with insurance of correct applying. In your case you can modify something in first set of tests and forget to modify it in second set, so you will run different tests and it could lead to hard-catch errors.

File: reverse.js (hunk 2 above)
Comment (olvnikon) [reply reverse.js]: Agree. Because here you should have a number according to the regular expression. But you should update it though because right now it doesn't work properly. Take a look at these cases:
```
/\d/.test("2") -> true
/\d/.test("2a") -> true
/\d/.test("dfkjhskjghs78sdhfjskdfhksjdfs2 #I&@*$@#U*(#@") -> true
```

File: reverse.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  else {
+
+    var arr = string.split(" ");
+    for (var k = 0; k < string.length;  k++) {
+      character = arr[k];
+
+      if (/\d/.test(character)){
+        stack.push(character);
+      } else if (character in operators) {
+        var b = parseInt(stack.pop());
```
Comment (olvnikon) [inline reverse.js]: Please, update names of variables because `a` and `b` are unclear names and you can forget what are they.

File: test.js (hunk 10 above)
Comment (olvnikon) [reply test.js]: Yes, you are right. Tests should be updated.
```
4 passing
6 failing
```

File: reverse.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (ciraelith) [reply reverse.js]: @olvnikon and why !string is not enough?

File: reverseE6.js (hunk 3 above)
Comment (ciraelith) [reply reverseE6.js]: That was a mistake of file :)

File: reverse.js (hunk 12 above)
Comment (ciraelith) [reply reverse.js]: These are mathematical variables corresponding to the calculating functions. They are widely-spread so there is nothing unclear about them. Like in those examples: https://www.w3schools.com/js/js_function_definition.asp

File: reverse.js (hunk 2 above)
Comment (ciraelith) [reply reverse.js]: I'll update it to /^\d+$/

File: reverse.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (olvnikon) [reply reverse.js]: @ciraelith yep, sorry. I just wanted to say that you could use exclamation marks to convert to boolean.

File: reverse.js (hunk 2 above)
Comment (olvnikon) [reply reverse.js]: Nope. It is still not valid.
`/\d+/.test("1a4") -> true`
Do you need help for this?

Comment (olvnikon) [reply reverse.js]: Yep, it looks valid. Use this.

File: reverse.js (hunk 12 above)
Comment (olvnikon) [reply reverse.js]: It is just an example. `foo`, `bar`, `baz` etc. are also very popular. :) In real projects, if you see something like `var a` you have no idea what is it. When you see a variable you should be able to understand what is it even w/o the right side of the evaluation.
```
var secondOperand = parseInt(stack.pop());
var firstOperand = parseInt(stack.pop());
```
Sometimes variables can have long names and it is ok. `applicationDefaultSettingsMapping` better than `mapping`.
@PinkaminaDianePie what do you think about this?

Comment (olvnikon) [reply reverse.js]: Ok. There are some cases when you can use these short names. For example, this one:
`"-": (a, b) => a - b,`
This is okay because there is nothing else except this variables. In such cases go ahead with short names if they fit the situation.

File: JSCore06/reverseE6.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  };
+  let operators = Object.assign({}, baseOperators, extraOperators);
+
+  const isProperString = /^[0-9*+ ().,-]+$/.test(string);
+  if (!string || !isProperString ) {
+    throw new Error ("Please put an equation to solve in a form of string with spaces");
+  } else {
+    let arr = string.split(" ");
+    for (let k = 0; k < string.length;  k++) {
+      let character = arr[k];
```
Comment (olvnikon) [inline JSCore06/reverseE6.js]: After you change the solution to reduce it will become `const`.

File: JSCore06/reverseE6.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      if (/^\d+$/.test(character)){
+        stack.push(character);
+      } else if (character in operators) {
+        let secondOperand = parseInt(stack.pop());
```
Comment (olvnikon) [inline JSCore06/reverseE6.js]: After you change the solution to reduce `secondOperand`, `firstOperand` and `value` will become `const`.

File: JSCore06/reverseE6.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        let firstOperand = parseInt(stack.pop());
+        let value = operators[character](firstOperand, secondOperand);
+        stack.push(value);
+      }
+    }
+    let result = stack[0];
```
Comment (olvnikon) [inline JSCore06/reverseE6.js]: It is `const`.

File: JSCore06/runner.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
+const string = process.argv.slice(2)[0];
```
Comment (olvnikon) [inline JSCore06/runner.js]: Extra operators don't work.
```
node runner.js "2 2 /"
\EducationalPlan\JSCore06\reverseE6.js:12
    throw new Error ("Please put an equation to solve in a form of string with spaces");
    ^

Error: Please put an equation to solve in a form of string with spaces
```

File: JSCore06/test.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
+const assert = require('assert');
+const reverseCalc = require('./reverseE6').reverseCalc;
+const reverseCalc5 = require('./reverse').reverseCalc;
+
+describe('reverseCalc, reverseCalc5', function() {
+    it('should properly add 2+2 during the test', function() {
```
Comment (olvnikon) [inline JSCore06/test.js]: You don't have to write `during the test` because it is obvious. Please, remove it from each test.

File: JSCore06/test.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+      const substraction = '4 2 -';
+        assert.equal(reverseCalc(substraction), 2);
+        assert.equal(reverseCalc5(substraction), 2);
+    });
+    it('should properly add 2+2+2 during the test', function() {
+      const addition = '2 2 2 + +';
+        assert.equal(reverseCalc(addition), 6);
+        assert.equal(reverseCalc5(addition), 6);
+    });
+    it('should properly multiply 4*4*4 during the test', function() {
```
Comment (olvnikon) [inline JSCore06/test.js]: The description of the test doesn't fit the test.
`const multiplying = '2 2 2 * *';`

File: JSCore06/test.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+      const multiplying = '2 2 2 * *';
+        assert.equal(reverseCalc(multiplying), 8);
+        assert.equal(reverseCalc5(multiplying), 8);
+    });
+    it('should properly substract 4-2 during the test', function() {
```
Comment (olvnikon) [inline JSCore06/test.js]: The description of the test doesn't fit the test. Looks like copy-paste.
`const substraction = '2 2 10 - -';`

File: JSCore06/reverse.js [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    throw new Error("Please put an equation to solve in a form of string with spaces");
+  } else {
+    var arr = string.split(" ");
+    for (var k = 0; k < string.length;  k++) {
+      character = arr[k];
+
+      if (/^\d+$/.test(character)){
+        stack.push(character);
+      } else if (character in operators) {
+        var secondOperand = parseInt(stack.pop());
```
[the remark this answers is not in the document]
Comment (PinkaminaDianePie) [reply JSCore06/reverse.js]: and we already discussed this 2 times before

File: JSCore06/reverseE6.js [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    if (stack.length > 1 || isNaN(stack)){
+      throw Error("Please put equation in correct format");
+    }
+      console.log("Your ES6 result is: " + result);
```
Comment (PinkaminaDianePie) [inline JSCore06/reverseE6.js]: Use [template strings](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/template_strings) to interpolate any data to string in es6

File: JSCore06/reverseE6.js [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+    "*": (a, b) => a * b,
+  };
+  let operators = Object.assign({}, baseOperators, extraOperators);
+
+  const isProperString = /^[0-9*+ ().,-]+$/.test(string);
+  if (!string || !isProperString ) {
+    throw new Error ("Please put an equation to solve in a form of string with spaces");
+  } else {
+    let arr = string.split(" ");
+    for (let k = 0; k < string.length;  k++) {
```
[the remark this answers is not in the document]
Comment (PinkaminaDianePie) [reply JSCore06/reverseE6.js]: It's better to use `reduce` in es5 version. In es6 you can use [for-of](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Statements/for...of) cycle

File: JSCore06/reverse.js [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+        var firstOperand = parseInt(stack.pop());
+        var value = operators[character](firstOperand, secondOperand);
+        stack.push(value);
+      }
```
Comment (olvnikon) [inline JSCore06/reverse.js]: We also need to add `else` statement which should throw an error because if it is not a number and an operator it is some incorrect character. For example:
```
node runner.js "( 2 2 +"
[ 4 ]
Your ES6 result is: 4
Your ES5 result is: 4
```
It works but it shouldn't.

File: JSCore06/reverse.js [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+  var operators = extraOperators
+  ? mergeOptions(baseOperators, extraOperators)
+  : baseOperators;
+
+  var isProperString = /^[0-9*+ ().,-]+$/.test(string);
+  if (!string || !isProperString ) {
+    throw new Error("Please put an equation to solve in a form of string with spaces");
+  } else {
+    var arr = string.split(" ");
+    for (var k = 0; k < string.length;  k++) {
```
Comment (olvnikon) [inline JSCore06/reverse.js]: You should take the length of the array.
arr = ["2", "2", "+"] -> length 3
string = "2 2 +" -> length 5

File: JSCore06/extraOperators.js [hunk 25]
```diff
@@ -0,0 +1,3 @@
+const extraOperators = {
+  "/": function(a,b) { return a/b }
```
[the remark this answers is not in the document]
Comment (ciraelith) [reply JSCore06/extraOperators.js]: Is semicolon after return in one-line function necessary?

[the remark this answers is not in the document]
Comment (olvnikon) [reply JSCore06/extraOperators.js]: Not after. Inside. Let's say if you don't have everything will work just fine but according to the style guide, you should add it. https://github.com/airbnb/javascript/blob/master/README.md#semicolons--required
`function(a, b) { return a / b; }`
